{
 "metadata": {
  "name": "QTrim"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Removal of artefacts and low quality nucleotides from Reference Assembly reads\n",
      "------------------------------------------------------------------------------\n",
      "\n",
      "* Read sets are searched for contamination by Illumina adapters and primers using cutadapt ver. 1.1.\n",
      "* Following removal of artefactual sequences, low quality basepairs are removed. (Here, nucleotides with a phred score less than 20 are below the\n",
      "quality threshold.)\n",
      "* Both 5 and 3 prime ends of the read are searched for low quality nucleotides and trimmed. After 1st low quality nucleotide is found, the sequence string of 15 nucleotides down-/upstream from that position is evaluated and trimmed if below phred threshold.\n",
      "* Reads that contain more than 5 nucleotides below phred threshold after 5 and 3 prime trimming are removed.\n",
      "* Only reads that exceed or are equal to 45 basepairs in length are retained.\n",
      "        ```\n",
      "        # Convert old Illumina quality encoding to Sanger\n",
      "        Convert.Convert(Qual)\n",
      "        # Initialize trimming tools\n",
      "        Trim = PhredTools.EndTrim(Seq, NewQual, 20)\n",
      "        # Trim 5-3 prime\n",
      "        Trim.FivePrime(15)\n",
      "        # Trim 3-5 prime\n",
      "        Trim.ThreePrime(15)\n",
      "        # Remove reads with > 5 poor quality nucleotides\n",
      "        Trim.IntraTrim(5)\n",
      "        # Remove reads that are < 45 bp long\n",
      "        Trim.MinLength(45)```\n",
      "\n",
      "* Trimming and pairing of paired reads is performed in one pass. Input and output are compressed (gzip) fastq files to keep\n",
      "storage space needed small."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import os, sys, glob, subprocess, gzip\n",
      "sys.path.append(os.path.abspath('../code/python/cutadapt-1.1/cutadapt/scripts/'))\n",
      "import cutadapt\n",
      "sys.path.append('../code/python/')\n",
      "import PhredTools\n",
      "\n",
      "\"\"\"\n",
      "RawDataDir is the path to the folders containing the raw read files.\n",
      "\"\"\"\n",
      "#RawDataDir = ('/mnt/pond/BGI/T6/RawData/')\n",
      "RawDataDir = raw_input('Type in path to raw data. Path must end in directory named RawData!:\\n')\n",
      "\n",
      "\"\"\"\n",
      "Create a sorted list of files to process\n",
      "\"\"\"\n",
      "#Iterate though the reference data files\n",
      "#RawReads = [file for file in glob.glob('%s%s' % (RawDataDir, 'D[CN][1-8]0*/*'))]\n",
      "\n",
      "# Iterate through all read-sets\n",
      "RawReads = [file for file in glob.glob('%s%s' % (RawDataDir, 'D[CN][1-8]*/*'))]\n",
      "RawReads.sort()\n",
      "\n",
      "\n",
      "\"\"\"\n",
      "Use cutadapt in a sub-process to remove artefactual sequences; low quality nucleotides\n",
      "are removed using PhredTools\n",
      "\"\"\"\n",
      "# Create and instance of the Phredscore converter\n",
      "Convert = PhredTools.Convert()\n",
      "\n",
      "#Define the strings necessary to call cutadapt as a subprocess of this script\n",
      "str0 = \"\"\"python -c '\n",
      "import sys\n",
      "sys.path.append(\"/home/bastodian/Dropbox/github/Dimensions/T6/code/python/cutadapt-1.1/cutadapt/scripts/\")\n",
      "import cutadapt\n",
      "cutadapt.main([\"-g\", \"AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT\", \"-b\", \"GATCGGAAGAGCACACGTCTGAACTCCAGTCAC\", \"-b\", \"ATCTCGTATGCCGTCTTCTGCTTG\", \\\"\"\"\"\n",
      "str1 = \"\"\"python -c '\n",
      "import sys\n",
      "sys.path.append(\"/home/bastodian/Dropbox/github/Dimensions/T6/code/python/cutadapt-1.1/cutadapt/scripts/\")\n",
      "import cutadapt\n",
      "cutadapt.main([\"-g\", \"AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTAGATCTCGGTGGTCGCCGTATCATT\", \"-b\", \"CAAGCAGAAGACGGCATACGAGAT\", \"-b\", \"GTGACTGGAGTTCAGACGTGTGCTCTTCCGAT\", \\\"\"\"\"\n",
      "str2 = \"\"\"\\\"])\\nsys.stdout.flush()'\"\"\"\n",
      "\n",
      "# Step though the files containing fwd reads\n",
      "for index in range(0,len(RawReads[:]),2):\n",
      "    # Pair fwd and reverse read files\n",
      "    Fwd = RawReads[index]\n",
      "    Rev = RawReads[index + 1]\n",
      "    assert '1.fq.gz' in Fwd\n",
      "    assert '2.fq.gz' in Rev\n",
      "    \n",
      "    if Fwd.rstrip('1.fq.gz') == Rev.rstrip('2.fq.gz'):\n",
      "        print Fwd, '\\n', Rev\n",
      "        # Create the directory to write trimmed read files into\n",
      "        OutDir = '/'.join(Fwd.split('/')[:-1]).replace(\"RawData\", \"TrimmedData\")\n",
      "        try:\n",
      "            os.makedirs(OutDir)\n",
      "        except Exception:\n",
      "            pass\n",
      "        # Start cutadapt as two separate subprocesses for Fwd and Rev\n",
      "        cmd1 = '%s%s%s' % (str0, Fwd, str2)\n",
      "        process0 = subprocess.Popen(cmd1, shell=True, stdout=subprocess.PIPE)\n",
      "        cmd2 = '%s%s%s' % (str1, Rev, str2)\n",
      "        process1 = subprocess.Popen(cmd2, shell=True, stdout=subprocess.PIPE)\n",
      "    else:\n",
      "        sys.exit('Files paired incorrectly... Quitting')\n",
      "    \n",
      "    # Open a gzipped file for storing the trimmed reads\n",
      "    OutFile0 = '%s%s' % (Fwd.rstrip(\"fq.gz\").replace(\"RawData\", \"TrimmedData\"), '.trimmed.fq.gz')\n",
      "    OutFile1 = '%s%s' % (Rev.rstrip(\"fq.gz\").replace(\"RawData\", \"TrimmedData\"), '.trimmed.fq.gz')\n",
      "    print OutFile0, '\\n', OutFile1\n",
      "    with gzip.open(OutFile0, 'wb') as TrimOut0:\n",
      "        with gzip.open(OutFile1, 'wb') as TrimOut1:\n",
      "            #Read through each file line by line as passed by cutadapt to stdout\n",
      "            while True:\n",
      "                # Retrieve output from cutadapt\n",
      "                Header0 = process0.stdout.readline().strip()\n",
      "                Seq0 = process0.stdout.readline().strip()\n",
      "                Spacer0 = process0.stdout.readline().strip()\n",
      "                Qual0 = process0.stdout.readline().strip()\n",
      "                NewQual0 = '%s' % Convert.Convert(Qual0)\n",
      "\n",
      "                Header1 = process1.stdout.readline().strip()\n",
      "                Seq1 = process1.stdout.readline().strip()\n",
      "                Spacer1 = process1.stdout.readline().strip()\n",
      "                Qual1 = process1.stdout.readline().strip()\n",
      "                NewQual1 = '%s' % Convert.Convert(Qual1)\n",
      "\n",
      "                # Trim the fwd reads using PhreadTools\n",
      "                Trim0 = PhredTools.EndTrim(Seq0, NewQual0, 20)\n",
      "                Trim0.FivePrime(15)\n",
      "                Trim0.ThreePrime(15)\n",
      "                Trim0.IntraTrim(5)\n",
      "                Trim0.MinLength(45)\n",
      "                \n",
      "                # Verify that we received a sequence; if yes move on to reverse reads\n",
      "                if None in Trim0.Retrieve():\n",
      "                    continue\n",
      "                else:\n",
      "                     # Trim the fwd reads using PhreadTools\n",
      "                    Trim1 = PhredTools.EndTrim(Seq1, NewQual1, 20)\n",
      "                    Trim1.FivePrime(15)\n",
      "                    Trim1.ThreePrime(15)\n",
      "                    Trim1.IntraTrim(5)\n",
      "                    Trim1.MinLength(45)\n",
      "                    \n",
      "                    # Make sure that we have a reverse read; if no, move on to net read pair\n",
      "                    if not None in Trim1.Retrieve():\n",
      "                        TrimOut0.write('%s\\n%s\\n%s\\n%s\\n' % (Header0, Trim0.Retrieve()[0], Spacer0, Trim0.Retrieve()[1]))\n",
      "                        TrimOut1.write('%s\\n%s\\n%s\\n%s\\n' % (Header1, Trim1.Retrieve()[0], Spacer1, Trim1.Retrieve()[1]))\n",
      "                    else:\n",
      "                        continue"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": "*"
    }
   ],
   "metadata": {}
  }
 ]
}